nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-parameter-contexts.js

```diff
@@ -0,0 +1,2092 @@
+                            }
+
+                            // toggle this block
+                            toggle(twist, list);
+
+                            // update the border if necessary
+                            updateReferencingComponentsBorder($('#parameter-referencing-components-container'));
+                        }).append(twist).append(title).append(count).appendTo(referencingPgReferencesContainer);
+
+                        // add the listing
+                        list.appendTo(referencingPgReferencesContainer);
+                    };
+
+                    // get process group and show name or UUID based on user permissions
+                    var pg = nfProcessGroup.get(key);
+                    if (!nfCommon.isUndefinedOrNull(pg)) {
+                        if (pg.permissions.canRead === true) {
+                            // create block for this process group
+                            createReferenceBlock(pg.component.name, groups);
+                        } else {
+                            // create block for this process group
+                            createReferenceBlock(key, groups);
+                        }
+                    } else {
+                        var processGroupName = key;
+
+                        // attempt to resolve the group name
+                        var breadcrumbs = nfNgBridge.injector.get('breadcrumbsCtrl').getBreadcrumbs();
+                        $.each(breadcrumbs, function (_, breadcrumbEntity) {
+                            if (breadcrumbEntity.id === key) {
+                                processGroupName = breadcrumbEntity.label;
+                                return false;
+                            }
+                        });
+                        // create block for this process group
+                        createReferenceBlock(processGroupName, groups);
+                    }
+                }
+            }
+        }
+    };
+
+    /**
+     * Sorts the specified entities based on the name.
+     *
+     * @param {object} a
+     * @param {object} b
+     * @returns {number}
+     */
+    var nameComparator = function (a, b) {
+        return a.component.name.localeCompare(b.component.name);
+    };
+
+    /**
+     * Adds a new parameter.
+     */
+    var addNewParameter = function () {
+        var parameterName = $.trim($('#parameter-name').val());
+
+        // ensure the parameter name is specified
+        if (parameterName !== '') {
+            var parameterGrid = $('#parameter-table').data('gridInstance');
+            var parameterData = parameterGrid.getData();
+
+            // ensure the parameter name is unique
+            var matchingParameter = null;
+            $.each(parameterData.getItems(), function (_, item) {
+                if (parameterName === item.name) {
+                    matchingParameter = item;
+                }
+            });
+
```

Comment:
General question... Are there any other validations that needs to happen when creating/updating parameters? Are there any characters that are not allowed?

Reply:
I see in the nfel.js we have `const parameterKeyRegex = /^[a-zA-Z0-9-_. ]+/;`. I will add a validation check for this when creating a new parameter.